Migrate LIFX to config entry per device

bdraco opened the pull request:
Since there were no tests for this integration, and we are doing a migration, it seemed prudent to get to 100% coverage.  The downside to that is this PR is a bit larger than I had hoped, but at least the majority of the changes by line count is the addition of tests.

## Breaking change

Each LIFX device now has its config entry. The existing entry will be migrated to separate entries automatically as each device is discovered, and will automatically be removed once all devices are found. Migrating to a config entry per device model allows Home Assistant to take care of discovery IP updates and retries, dramatically improving the integration's reliability.

LIFX can now be configured manually by IP address. DHCP, HomeKit, and native discovery are supported for updating IP addresses to ensure we never lose track of a device.

Advanced YAML configuration is deprecated and will be removed in a future release. Advanced YAML configuration is no longer necessary as the integration now fully utilizes [Network Configuration](https://www.home-assistant.io/integrations/network/), and can also function without working discovery.

## Proposed change

Since the LIFX integration currently has discovery reliability issues similar to tplink integration used to have, migrating to a config entry per device model (similar to what is used for other wifi lighting devices like `tplink`, `flux_led`, `wiz`, `nanoleaf`, `wled`, `shelly`, and `yeelight`) which allows Home Assistant to take care of discovery ip updates and retries similar to how we did this for tplink in #56701.

- [x] Lots of tests
- [x] Move the connector into its own module called `aiolifx_connector`, and add a PR for the upstream lib https://github.com/frawau/aiolifx/pull/54 in case they find it useful there as well

Coverage is now 100%
```
---------- coverage: platform darwin, python 3.10.5-final-0 ----------
Name                                           Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------
homeassistant/components/lifx/__init__.py         89      0   100%
homeassistant/components/lifx/config_flow.py     122      0   100%
homeassistant/components/lifx/const.py             8      0   100%
homeassistant/components/lifx/coordinator.py      70      0   100%
homeassistant/components/lifx/discovery.py        33      0   100%
homeassistant/components/lifx/light.py           211      0   100%
homeassistant/components/lifx/manager.py          76      0   100%
homeassistant/components/lifx/migration.py        31      0   100%
homeassistant/components/lifx/util.py             83      0   100%
----------------------------------------------------------------------------
TOTAL                                            723      0   100%



```
## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue:  fixes #74246  fixes #74267 fixes #74002 fixes #74027 fixes #74021 fixes #72827 closes #74290
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23261

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Migrate LIFX to config entry per device

bdraco committed:
tweaks

bdraco committed:
fixes

bdraco committed:
tweaks

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
tidy up

bdraco committed:
fixes

bdraco commented on homeassistant/components/lifx/util.py:
```diff
@@ -0,0 +1,165 @@
+        return self.message
+
+
+class LIFXConnection:
+    """Manage a connection to a LIFX device."""
+
+    def __init__(self, host: str, mac: str) -> None:
+        """Init the connection."""
+        self.host = host
+        self.mac = mac
+        self.device: Light | None = None
+        self.transport: asyncio.DatagramTransport | None = None
+
+    async def async_setup(self) -> None:
+        """Ensure we are connected."""
+        loop = asyncio.get_running_loop()
+        transport_proto = await loop.create_datagram_endpoint(
+            lambda: Light(loop, self.mac, self.host),
+            remote_addr=(self.host, UDP_BROADCAST_PORT),
+        )
+        self.transport = cast(asyncio.DatagramTransport, transport_proto[0])
+        self.device = cast(Light, transport_proto[1])
+
+    def async_stop(self) -> None:
+        """Close the transport."""
+        assert self.transport is not None
+        self.transport.close()
```
TODO: Move the connector into the upstream lib

bdraco committed:
adjust

bdraco committed:
cleanups

bdraco committed:
cleanups

Djelibeybi commented:
This will break a lot of people's use of LIFX with Home Assistant. Your discovery interval is about 14 minutes and 50 seconds longer than people would like, because they rely on Home Assistant discovering LIFXbulbs that get physically powered off and on on a regular basis.

Djelibeybi commented:
That's why I backed out my version of this refactor and went back to a single config entry for all bulbs.

bdraco commented:
> This will break a lot of people's use of LIFX with Home Assistant. Your discovery interval is about 14 minutes and 50 seconds longer than people would like, because they rely on Home Assistant discovering LIFXbulbs that get physically powered off and on on a regular basis.

It won't actually. That's only the integration discovery interval.

DHCP and HomeKit discovery still happen immediately and config entry setup retry will trigger the moment the DHCP, HomeKit, or integration discovery finds a device.  Also normal retry will try to reconnect as well.

bdraco committed:
use testing lib for development

bdraco committed:
move details to upstream

bdraco commented on homeassistant/components/lifx/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/lifx",
-  "requirements": ["aiolifx==0.8.1", "aiolifx_effects==0.2.2"],
+  "requirements": ["haaiolifx==0.8.1", "aiolifx_effects==0.2.2"],
```
Temporary for easy testing until upstream PR is merged

bdraco committed:
start with tplink tests

bdraco committed:
tests part 1

bdraco committed:
tests part 1

bdraco committed:
cleanups

bdraco committed:
naming is hard

bdraco committed:
tweaks

bdraco committed:
fixes

bdraco committed:
cleanups

bdraco committed:
cleanups

bdraco committed:
merge tests

bdraco committed:
color temp bulb

bdraco committed:
more mocking

bdraco committed:
add cover

bdraco committed:
fix tests

bdraco committed:
delete unreachable code

bdraco committed:
delete unreachable code

bdraco committed:
tweaks

bdraco committed:
effect tests

bdraco committed:
fixes

bdraco committed:
adjust

bdraco committed:
remove unused code

bdraco committed:
has tests now

Djelibeybi commented:
Once you've happy with your progress, I think I'll take a shot at my dream refactor: replacing aiolifx with Photons. I've got a nearly complete custom integration using Photons already, so I'm optimistic if we combine forces, we'll end up with something both of us can support.

Djelibeybi commented:
Also, I haven't had a chance to look at your code yet because I've been running errands, but a reminder that the bulb's actual MAC address may be different to what aiolifx says it is. The value reported by aiolifx is its serial, which is different from firmware (3,70) and later.

bdraco commented:
> Also, I haven't had a chance to look at your code yet because I've been running errands, but a reminder that the bulb's actual MAC address may be different to what aiolifx says it is. The value reported by aiolifx is its serial, which is different from firmware (3,70) and later.

I think I've accounted for that quirk
https://github.com/home-assistant/core/pull/74316/files#diff-750013b3a6e79bea217a915ecdd8c433defd5ac436403a89ae4731347a098db0R124-R129

https://github.com/home-assistant/core/pull/74316/files#diff-925a26b942e34acee8ce3c6e1bab85fe954d6c6251657e2ea4c31b2e55edf2feR59-R70

bdraco committed:
add light strip zone update failure coverage

bdraco committed:
tweak

bdraco committed:
tweak

bdraco committed:
tweaks

bdraco committed:
fix set color

bdraco committed:
fix effect not being displayed in the UI

bdraco committed:
avoid 10s warning

bdraco commented:
@Djelibeybi Should be ready for testing now.  I haven't been able to break it but I only have 3 devices

bdraco committed:
coverage for get_version failure

bdraco committed:
version was missing

Djelibeybi commented:
@bdraco will take it for a spin now. As of a month or two ago, I now have at least one of every type of LIFX device active on the network. :) Let's see how much I hate configuring 60 individual devices.

bdraco committed:
cleanup discovery

Djelibeybi commented:
<img width="960" alt="integrations-page-madness" src="https://user-images.githubusercontent.com/103232/176986258-444e8d58-8351-4b8e-bb97-c05d1c217b1f.png">

Feature request: a method of selecting multiple integrations and mass-enabling/configuring all of them. Let's start the clickfest.

bdraco commented:
Probably need some more test coverage for the weird mac off by once case but I'll probably have to do that tomorrow as its getting really late here

Djelibeybi commented:
What's the best way to provide feedback? Directly here?

bdraco committed:
fixes for mac address

bdraco commented:
> What's the best way to provide feedback? Directly here?

Yes

Djelibeybi commented:
Discovery without an interface connected to the LIFX subnet found about half of the devices, with registration being somewhat slow. I know this because I forgot my test rig wasn't directly connected to my LIFX subnet these days. 😂  Have just enabled the directly attached interface to see if it finds the rest and if onboarding is faster.

Djelibeybi commented:
When I enable the directly connected interface, the LIFX integration discovery finds stuff but Home Assistant seems to get stuck on "LIFX integration is starting" -- no debug log output is occuring, so I don't know if it's validating the existing 30-odd devices I configured already.

bdraco commented:
> Edited to add: LIFX startup is not finishing. I'll take a look to see if I can work out why.

This logic looks a little fragile.  I wonder if its getting stuck here https://github.com/frawau/aiolifx/blob/8d5af5de9ddddd204f5f75d3fc87a5b33dbfd367/aiolifx/aiolifx.py#L297 and never getting the callback so it never times out

Djelibeybi commented:
Reviewing the new discovery logic now.

bdraco commented:
I tried unplugging on of them to replicate the stall on startup but no luck.  Everything starts up normally and the one I unplugged shows in a retry state
<img width="302" alt="Screen Shot 2022-07-01 at 23 53 57" src="https://user-images.githubusercontent.com/663432/176987039-15d5ee29-89f3-4938-a5cf-fade9a0b19b6.png">

Djelibeybi commented:
I haven't finished discovering/configuring all of mine yet. It shows the unconfigured ones on the integration page, but config flow doesn't work.

Djelibeybi commented:
I started with a fresh install, I didn't migrate an existing one.

bdraco commented:
> It shows the unconfigured ones on the integration page, but config flow doesn't work.

Can you give some more detail on what doesn't work?

Djelibeybi commented:
I didn't give it long enough, to be honest. The interface was very slow because Home Assistant hadn't finished startup, so it was too difficult to work with.

bdraco commented:
> I'll take a look to see if I can work out why.

That would be great.  If it is the underlying library (I can't think of what else it could be), it might explain some of the general unreliability as well.

Djelibeybi commented:
> That would be great. If it is the underlying library (I can't think of what else it could be), it might explain some of the general unreliability as well.

Given how significant this refactor is, would you be open to replacing `aiolifx` with Photons? I suspect I have most of the required code already written for my own integration. Will just need some tweaking.

bdraco commented:
> Given how significant this refactor is, would you be open to replacing `aiolifx` with Photons? I suspect I have most of the required code already written for my own integration. Will just need some tweaking.

I think that would be fine, but we shouldn't do it in this PR. It should be a followup as this one is already much larger than I generally like.

Djelibeybi commented:
I wonder if it makes more sense to switch from `aiolifx` to Photons first, then do this refactor to individual config entries? Let me noodle on that while you go to bed already. 😉

amelchio commented:
Regarding "general unreliability", note that LIFX bulbs are just incredibly finicky. Full reliability is probably a pipe dream.

But it's good to see the LIFX implementation catch up to current standards 👍

bdraco commented:
> > Edited to add: LIFX startup is not finishing. I'll take a look to see if I can work out why.
> 
> 
> 
> This logic looks a little fragile.  I wonder if its getting stuck here https://github.com/frawau/aiolifx/blob/8d5af5de9ddddd204f5f75d3fc87a5b33dbfd367/aiolifx/aiolifx.py#L297 and never getting the callback so it never times out

Can you still replicate the stall?  If so I'll wrap the waiter in an async timeout to be sure we handle timeout if the underlying library fails.   

Also the timeouts probably should be adjusted to back off a bit more. If the wifi is very latent, sending a request every second can make things worse as it ends up with a large retry queue. We found the wiz bulbs were more reliable with a larger interval and longer timeout

Djelibeybi commented:
> Can you still replicate the stall?

Yep, it's consistent for any new/fresh install, if the LIFX integration can detect devices. I can't get config flow to get past the first screen, with the logs just saying "Setting up LIFX" and then telling me how long it's taken to finish setup (which never happens).

Djelibeybi commented:
For reference, here are the packet gap timings from Photons:

```yaml
# These options are all optional, below are the defaults
      gaps:
        # Some messages, like that used for discovery has an "unbound" number of results
        # We stop waiting for results this gap plus 0.05 seconds before giving up waiting
        # for more results
        # In practise, with the default value, discovery will take a minimum 0.45 seconds to complete
        gap_between_results: 0.4

        # If we need both an acknowledgement and a result from a device, we wait this many seconds
        # after getting an acknowledgement before retrying a message if we haven't got a reply yet
        gap_between_ack_and_res: 0.2

        # This is the backoff times for retrying messages.
        # So with this setting, we retry at 0.2, 0.3, 0.4, 0.5, 0.7, 0.9, 1.1, 2.1, 3.1, 4.1, 5.1, 6.1, etc
        timeouts: [[0.2, 0.2], [0.1, 0.5], [0.2, 1], [1, 5]]
```

bdraco commented:
I wonder if the event reuse in the waiter is the cause of some of the unreliability.  I'm going to write a new waiter that uses a future and has an over all timeout.  That way we can throw exceptions for the silver requirement of raising HomeAssistantError on service call failure

bdraco commented:
> Yep, it's consistent for any new/fresh install, if the LIFX integration can detect devices. I can't get config flow to get past the first screen, with the logs just saying "Setting up LIFX" and then telling me how long it's taken to finish setup (which never happens).

Can you get to here when starting a new flow manually via add integration?
<img width="484" alt="Screen Shot 2022-07-02 at 09 13 19" src="https://user-images.githubusercontent.com/663432/177004407-294b56c4-687d-4e48-aa1e-ff2c2fe5f8fd.png">

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

Djelibeybi commented:
Yes, I can get to entering an IP address manually. If I provide an IP address, it results in a blank dialog and the config entry does not appear to be created, but no errors appear in the logs.

bdraco committed:
use a future instead

bdraco committed:
adjust mocking

bdraco commented:
> Yes, I can get to entering an IP address manually. If I provide an IP address, it results in a blank dialog and the config entry does not appear to be created, but no errors appear in the logs.

That's really odd.   Can you post a screen shot?

bdraco committed:
fix tests

bdraco committed:
fix mocks

bdraco committed:
coverage

bdraco committed:
failure test

bdraco committed:
failure test

bdraco committed:
merge tweaks

bdraco committed:
do not repoll the bulb since the ack sends back the new color anyways

Djelibeybi commented:
> That's really odd. Can you post a screen shot?

I can't reproduce it with the latest commits. Now I get this error if I enter an IP address manually:

```log
2022-07-03 01:59:14 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/Users/amiller/.pyenv/versions/venv-hass-dev/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/Users/amiller/.pyenv/versions/venv-hass-dev/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/Users/amiller/.pyenv/versions/venv-hass-dev/lib/python3.10/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/ban.py", line 79, in ban_middleware
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/auth.py", line 236, in auth_middleware
    return await handler(request)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/view.py", line 136, in handle
    result = await result
  File "/Users/amiller/Developer/hass/src/homeassistant/components/config/config_entries.py", line 177, in post
    return await super().post(request, flow_id)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/Users/amiller/Developer/hass/src/homeassistant/helpers/data_entry_flow.py", line 109, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 277, in async_configure
    result = await self._async_handle_step(
  File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 359, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/config_flow.py", line 100, in async_step_user
    device := await self._async_try_connect(host, raise_on_progress=True)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/config_flow.py", line 182, in _async_try_connect
    await connection.async_setup()
  File "/Users/amiller/.pyenv/versions/venv-hass-dev/lib/python3.10/site-packages/aiolifx/connection.py", line 42, in async_setup
    self.transport, self.device = await loop.create_datagram_endpoint(
  File "/Users/amiller/.pyenv/versions/3.10.4/lib/python3.10/asyncio/base_events.py", line 1304, in create_datagram_endpoint
    infos = await self._ensure_resolved(
  File "/Users/amiller/.pyenv/versions/3.10.4/lib/python3.10/asyncio/base_events.py", line 1404, in _ensure_resolved
    return await loop.getaddrinfo(host, port, family=family, type=type,
  File "/Users/amiller/.pyenv/versions/3.10.4/lib/python3.10/asyncio/base_events.py", line 860, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/amiller/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/amiller/.pyenv/versions/3.10.4/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

bdraco committed:
tweaks to reduce load on the device

bdraco committed:
tweaks to reduce load on the device

Djelibeybi commented:
Just a note: [the `State` reply to any `Set` message set to a bulb that changes the bulb's visible appearance is almost always the _previous_ state of the bulb](https://lan.developer.lifx.com/docs/changing-a-device), i.e. before the change. 

It's better to send a `Set` message with `ack_required=1` and `res_required=0` so you don't get the State response, then send a `GetColor` message afterwards to get the actual current state of the bulb.

This is something that Photons does natively but `aiolifx` appears to be unaware of. I'm assuming it's the reason for the very inefficient and convaluted `set_state` method used to transform the bulbs, when all of that can and should be done in with a single `SetWaveformOptional` message.

bdraco committed:
tweaks for light strips

bdraco committed:
reduce load on light strips

bdraco commented:
> I can't reproduce it with the latest commits. Now I get this error if I enter an IP address manually:

Did you have a typo in the ip address?  Its trying to resolve a name there.

Djelibeybi commented:
Yep, must have. Tried it again and got the blank box:

<img width="401" alt="Screen Shot 2022-07-03 at 2 48 08 am" src="https://user-images.githubusercontent.com/103232/177009284-c4b36165-eae5-45e8-961f-3b26a785a099.png">

There is absolutely nothing in the logs and I have `debug` enabled for `homeassistant.components.lifx`

bdraco committed:
handle the stale state case

bdraco commented:
> Just a note: [the `State` reply to any `Set` message set to a bulb that changes the bulb's visible appearance is almost always the _previous_ state of the bulb](https://lan.developer.lifx.com/docs/changing-a-device), i.e. before the change.
> 
That's lame

Handled in [home-assistant/core@`4a36530` (#74316)](https://github.com/home-assistant/core/pull/74316/commits/4a3653073ced3d1ff4ca36c8b6dca8a2583a4cdd)

bdraco committed:
add resolution failure catch

bdraco commented:
I added a check for the dns resolution failure in https://github.com/home-assistant/core/pull/74316/commits/6123d82bb1f13695b72482ea70dd42b902e766e5

bdraco committed:
fix raise_on_progress

bdraco commented:
> Yep, must have. Tried it again and got the blank box:
> 
> <img alt="Screen Shot 2022-07-03 at 2 48 08 am" width="401" src="https://user-images.githubusercontent.com/103232/177009284-c4b36165-eae5-45e8-961f-3b26a785a099.png">
> 
> There is absolutely nothing in the logs and I have `debug` enabled for `homeassistant.components.lifx`

raise_on_progress needed to be set to False in that path fixed in 1dd404922d19c4e27ad5161fd96273102a9f13ed

bdraco committed:
tweaks

bdraco committed:
coverage

bdraco committed:
tweaks

bdraco commented:
I need to write some more tests for light.py with all the refactoring, but everything appears to be working great with my devices now

bdraco committed:
small cleanups

bdraco committed:
more fixes

bdraco commented:
The light strips need a lot of coverage for the zones as that code path is a big complex

Djelibeybi commented:
> everything appears to be working great with my devices now

It's finally configuring devices for me, so that's a win. I'm going to switch to `dev`, let it populate all 60 devices, then switch back to this PR so it can convert them, instead of having to manually configure all 60. 😏

Djelibeybi commented:
It's gotten itself stuck trying to migrate 60 bulbs from legacy to config entries. Each attempt just takes too long and fetch is failing for most of the bulbs. It's probably worth doing the migration in batches of 5 at a time or so.

bdraco commented:
> It's gotten itself stuck trying to migrate 60 bulbs from legacy to config entries. Each attempt just takes too long and fetch is failing for most of the bulbs. It's probably worth doing the migration in batches of 5 at a time or so.

Is it just overwhelmed? Does it eventually finish with a lot of the migrations failing?

Djelibeybi commented:
> Does it eventually finish with a lot of the migrations failing?

I think "gives up" is a better description than "finish", but yes, all the migrations failed and it left me with broken config entries that can't start and all the bulbs rediscovered individually again.

I'm not sure what the response was from the TP-Link change, but I have to say, I'm not a fan of individual config entries. I've lost the ability to see all my LIFX devices/entities or how many it's detected or which ones are offline. The user experience for someone with a large fleet of bulbs is (in my opinion, subjectively) quite poor, to the point where I'm likely to use my custom integration in favour of this experience.

This should definitely get as long a beta period as possible before being merged, because folks have all sorts of usage patterns which previous refactors broke significantly enough to negatively impact quality of life for some people.

bdraco commented:
I think it's always going to have trouble with 60 bulb with the current design. Well it's always going to have trouble doing that many.  If we move the migration code into the legacy entry setup we can raise ConfigEntryNotReady and try again if the first attempt fails so they don't have to restart. It might take a minute or two to get everything migrated but that seems like a minor inconvenience

bdraco commented:
> I think "gives up" is a better description than "finish", but yes, all the migrations failed and it left me with broken config entries that can't start and all the bulbs rediscovered individually again.

We are never going to be able to discover them all right away at startup if there are many of them since there is too much competition for resources.  I'll work on the approach I outlined above to accommodate larger setups

> I'm not sure what the response was from the TP-Link change, but I have to say, I'm not a fan of individual config entries. I've lost the ability to see all my LIFX devices/entities or how many it's detected or which ones are offline. The user experience for someone with a large fleet of bulbs is (in my opinion, subjectively) quite poor, to the point where I'm likely to use my custom integration in favour of this experience.

That is a common complaint about the UI when there are a lot of config entries.  I agree that it could be improved, and there has been a lot of focus on improving the frontend and streamlining the experience this year.  Deficiencies in the frontend should be fixed, and we should not design the backend to accommodate them instead.

bdraco committed:
retry migration

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco commented:
The migration is also unreliable because the version on `dev` has the wrong mac address because the call to `get_hostfirmware` was removed at some point so it never hits the logic to change the last byte

bdraco commented:
We end up with duplication migrations because it doesn't know if its been migrated or not since it does it by mac addresses which leads to the failure state

bdraco committed:
fixes

bdraco commented:
I think I need to build a helper that will match a mac address off by one either way

bdraco committed:
fixes

Djelibeybi commented:
Why not ignore MAC mac address and just use the serial as the identifier, which is the same regardless of firmware version? The only time Home Assistant uses the real MAC address is to create the connections entry in `DeviceInfo`.

bdraco commented:
> Why not ignore MAC mac address and just use the serial as the identifier, which is the same regardless of firmware version? 

I'm confused how its the same pre and post 3.70 if we need the get real mac logic.  Is there some documentation that explains what they actually changed between versions?

> The only time Home Assistant uses the real MAC address is to create the connections entry in `DeviceInfo`.

Djelibeybi commented:
> I'm confused how its the same pre and post 3.70

This is an `aiolifx` bug/problem: what it stores in `Device.mac_addr` is the serial, not the MAC address so I created the conversion function to get the real MAC for `DeviceInfo` so that Home Assistant could create valid connections with things like the UniFi integration.

bdraco commented:
Probably need to make all the unique ids the serial number and then when we discover via dhcp we can check if within one digit

bdraco commented:
I'm going to refactor a bit to change everything to serial number or mac so it's not as confusing as they are used interchangeably in the lib

Djelibeybi commented:
> Probably need to make all the unique ids the serial number and then when we discover via dhcp we can check if within one digit

The unique id _is_ the serial number. :)

bdraco commented:
I'm going to wrap all the calls made to the lib directly from light.py as well on to the coordinator class so it's easier to change out the underlying library later

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
move get/set calls to the coordinator

bdraco committed:
naming

bdraco commented:
> > Probably need to make all the unique ids the serial number and then when we discover via dhcp we can check if within one digit
> 
> The unique id _is_ the serial number. :)

I used the mac address for the config entry unique id because I wanted to match it to the DHCP address.  I'm going to switch that to be the serial number and create a matching function

bdraco committed:
cleanup

bdraco committed:
add comment about device.mac_addr

bdraco committed:
add comment about device.mac_addr

bdraco committed:
naming

bdraco committed:
fixes

bdraco committed:
coverage

bdraco committed:
fixes

bdraco committed:
merge

bdraco commented:
Well I can't do one migration at a time because I can't await it because it will deadlock

bdraco committed:
fix

Djelibeybi commented:
The lastest code migrated all but one entry, which resulted in a key error:

```plain
Logger: homeassistant.config_entries
Source: components/lifx/migration.py:37
First occurred: 11:22:01 AM (1 occurrences)
Last logged: 11:22:01 AM

Error setting up entry LIFX for lifx
Traceback (most recent call last):
  File "/Users/amiller/Developer/hass/src/homeassistant/config_entries.py", line 353, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/__init__.py", line 113, in async_setup_entry
    await async_legacy_migration(hass, entry)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/__init__.py", line 81, in async_legacy_migration
    missing_discovery_count = await async_migrate_legacy_entries(
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/migration.py", line 37, in async_migrate_legacy_entries
    async_init_discovery_flow(hass, discovered_hosts_by_serial[serial], serial)
KeyError: 'd0:73:d5:26:3f:b4'
```

Though, that device is responding to packets and it has the same MAC and serial, so 🤷🏻

Djelibeybi commented:
Restarting cleared things and it migrated and now they're all migrated.

bdraco committed:
retry migration if it fails the first time

bdraco committed:
retry migration if it fails the first time

bdraco committed:
adjust calls

bdraco committed:
adjsut

bdraco commented:
> The lastest code migrated all but one entry, which resulted in a key error:
> 
> ```
> Logger: homeassistant.config_entries
> Source: components/lifx/migration.py:37
> First occurred: 11:22:01 AM (1 occurrences)
> Last logged: 11:22:01 AM
> 
> Error setting up entry LIFX for lifx
> Traceback (most recent call last):
>   File "/Users/amiller/Developer/hass/src/homeassistant/config_entries.py", line 353, in async_setup
>     result = await component.async_setup_entry(hass, self)
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/__init__.py", line 113, in async_setup_entry
>     await async_legacy_migration(hass, entry)
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/__init__.py", line 81, in async_legacy_migration
>     missing_discovery_count = await async_migrate_legacy_entries(
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/migration.py", line 37, in async_migrate_legacy_entries
>     async_init_discovery_flow(hass, discovered_hosts_by_serial[serial], serial)
> KeyError: 'd0:73:d5:26:3f:b4'
> ```
> 
> Though, that device is responding to packets and it has the same MAC and serial, so 🤷🏻

I fixed that in one of the previous passes. I had a reversed conditional

bdraco commented:
I worked around the deadlock.  The migration will run every time there is a successful discovery until it is complete.  So if the first time it doesn't find all the bulbs it will keep trying until it finds them all or they delete the legacy config entry.

bdraco committed:
break it up

bdraco committed:
fixes for dhcp discovery

bdraco commented:
With all the changes back down to 95% test cover.

Will retest everything then write the missing tests

bdraco commented:
I broke something with setting the color zones when I moved the code to the coordinator.  But I didn't have tests for that yet so no suprise

bdraco commented:
> I broke something with setting the color zones when I moved the code to the coordinator. But I didn't have tests for that yet so no suprise

Actually it works fine.  I just had the brightness set to 0 🤦

bdraco commented:
All the testing looks good.

Writing some more tests but will have to sleep soon

Djelibeybi commented:
Sleep first, but the lastest code stalls on migration again. The LIFX integration doesn't finish setup. I get a whole bunch of logs, up to "confirming discovery" and then ... nothing.

bdraco committed:
fix dhcp discovery

bdraco committed:
account for I/O-cpu at startup

bdraco committed:
account for I/O-cpu at startup

bdraco committed:
cooldown discovery

bdraco commented:
> Sleep first, but the lastest code stalls on migration again. The LIFX integration doesn't finish setup. I get a whole bunch of logs, up to "confirming discovery" and then ... nothing.

I changed it to be an untracked task so it doesn't block startup as its going to take too long with 60 of them or its heading the deadlock thats fixed in #73806

bdraco committed:
merge

bdraco committed:
typo

bdraco committed:
reduce branching

bdraco committed:
reduce branching

bdraco committed:
reduce branching

bdraco committed:
reduce branching

Djelibeybi commented:
I got this error during the migration, but it seems to have completed successfully:

```python
Logger: homeassistant
Source: components/lifx/util.py:114
First occurred: 1:44:17 PM (57 occurrences)
Last logged: 1:44:17 PM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 222, in async_init
    flow, result = await task
  File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 249, in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data, init_done)
  File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 359, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/config_flow.py", line 50, in async_step_dhcp
    if entry.unique_id and mac_matches_serial_number(mac, entry.unique_id):
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 139, in mac_matches_serial_number
    or _get_mac_offset(serial_number, 1) == formatted_mac
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 114, in _get_mac_offset
    octets = [int(octet, 16) for octet in mac_addr.split(":")]
  File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 114, in <listcomp>
    octets = [int(octet, 16) for octet in mac_addr.split(":")]
ValueError: invalid literal for int() with base 16: 'lifx'
```

bdraco commented:
> I got this error during the migration, but it seems to have completed successfully:
> 
> ```python
> Logger: homeassistant
> Source: components/lifx/util.py:114
> First occurred: 1:44:17 PM (57 occurrences)
> Last logged: 1:44:17 PM
> 
> Error doing job: Task exception was never retrieved
> Traceback (most recent call last):
>   File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 222, in async_init
>     flow, result = await task
>   File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 249, in _async_init
>     result = await self._async_handle_step(flow, flow.init_step, data, init_done)
>   File "/Users/amiller/Developer/hass/src/homeassistant/data_entry_flow.py", line 359, in _async_handle_step
>     result: FlowResult = await getattr(flow, method)(user_input)
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/config_flow.py", line 50, in async_step_dhcp
>     if entry.unique_id and mac_matches_serial_number(mac, entry.unique_id):
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 139, in mac_matches_serial_number
>     or _get_mac_offset(serial_number, 1) == formatted_mac
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 114, in _get_mac_offset
>     octets = [int(octet, 16) for octet in mac_addr.split(":")]
>   File "/Users/amiller/Developer/hass/src/homeassistant/components/lifx/util.py", line 114, in <listcomp>
>     octets = [int(octet, 16) for octet in mac_addr.split(":")]
> ValueError: invalid literal for int() with base 16: 'lifx'
> ```

Its trying to process the legacy entry it needs an exclude

bdraco committed:
add legacy guard

bdraco commented:
fixed in 4d5ae0bfdfee5f12856c81eafdf34a0ce76acffa

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco commented on tests/components/lifx/test_init.py:
```diff
@@ -0,0 +1,146 @@
+    bulb = _mocked_failing_bulb()
+
+    class MockLifxConnectonDnsError:
+        """Mock lifx discovery."""
```
```suggestion
        """Mock lifx connection with a dns error."""
```

bdraco committed:
Update tests/components/lifx/test_init.py

bdraco committed:
add fast path tests

bdraco committed:
test for zones

bdraco committed:
add color temp tests for color + color temp bulb

bdraco committed:
add tests for a brightness only device

bdraco committed:
remove unreachable code

bdraco committed:
add tests for transitions

bdraco committed:
add test for set_infrared

bdraco committed:
fix off transitions

bdraco committed:
cover out of sync state case

bdraco committed:
migrate to package

bdraco committed:
tweaks

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
put the delay back in that was removed when testing

bdraco committed:
Merge branch 'lifx_migrate_config_entry_per_device' of github.com:bdraco/home-assistant into lifx_migrate_config_entry_per_device

Djelibeybi commented:
Would it be too much to add an "Identify" button entity to each device as part of this PR? I have a bunch of entities waiting for the more specific bulb types (particularly for HEV support) so it would be great to see the best way of handling multiple devices with the refactor.

bdraco commented:
> Would it be too much to add an "Identify" button entity to each device as part of this PR? I have a bunch of entities waiting for the more specific bulb types (particularly for HEV support) so it would be great to see the best way of handling multiple devices with the refactor.

Should be pretty easy to do, but I'll do that in a followup as I don't want to make this PR any larger.

Djelibeybi commented:
I hadn't done regression tests with your latest changes but we're back to pre-2022.5 for user experience: the `LIFXConnection` works even if LIFX discovery doesn't find the bulb, because it'll use unicast to connect from Home Assistant via whatever interface is enabled. Assuming that routing and firewall rules permit, the devices will continue to work without a locally enabled interface on the same subnet.

So... win?

bdraco commented:
> > Would it be too much to add an "Identify" button entity to each device as part of this PR? I have a bunch of entities waiting for the more specific bulb types (particularly for HEV support) so it would be great to see the best way of handling multiple devices with the refactor.
> 
> Should be pretty easy to do, but I'll do that in a followup as I don't want to make this PR any larger.

Don't hesitate to remind me after this merges in case I forget.  We might have to wait a bit for another core dev to take a look since the release on wed will eat up a lot of time and the summer holiday always makes everything take a lot longer.

Djelibeybi commented:
Sure, no worries. I'm currently toying with the idea of creating a drop-in replacement for aiolifx using Photons so that more projects could take advantage of it. The actual Photons developer always wanted to do it when he was still at LIFX, but doesn't have the time (or frankly the enthusiam) to do it now.

I think letting folks get used to this change first before introducing new stuff is probably a good idea too.

Djelibeybi commented:
FYI, it took four restarts to get Home Assistant to migrate all my LIFX bulbs to config entries. I still have a "LIFX" entry, but it's empty. I'm assuming it'll disappear eventually.

Djelibeybi commented on homeassistant/components/lifx/util.py:
```diff
@@ -0,0 +1,162 @@
+    async with async_timeout.timeout(OVERALL_TIMEOUT):
+        result = await future
+
+    _LOGGER.debug("LIFX command result: %s", result)
```
This is extremely noisy and I'm not convinced it's anything but noise for troubleshooting post-release. We probably don't need the entire response object in the logs for that.

bdraco replied:
removed via [`35e965a` (#74316)](https://github.com/home-assistant/core/pull/74316/commits/35e965a046870b64b3d307980f540cc3587980d2)

bdraco commented on homeassistant/components/lifx/util.py:
```diff
@@ -0,0 +1,162 @@
+    async with async_timeout.timeout(OVERALL_TIMEOUT):
+        result = await future
+
+    _LOGGER.debug("LIFX command result: %s", result)
```
```suggestion
```

bdraco committed:
Update homeassistant/components/lifx/util.py

bdraco commented:
I'm up to 8 devices with 5 more on the way.

Djelibeybi commented:
> We could use a shorter discovery interval when we know there are still devices that need migration

This is what `aiolifx` and Photons do, but in slightly different ways. Essentially, they start with short (1-1.5 second) broadcast timeouts and increase the interval between discoveries as it increases the discovery timeout so that it ramps down from the fast responders to the slow ones.

Djelibeybi commented:
> I'm up to 8 devices with 5 more on the way.

I have 60 devices _active_ on the network and more than what is reasonable sitting on my desk and in boxes and on shelves, etc. I'm sure I have at least another 5 LIFX Minis and a bunch of Z strips I can't find the controllers for.

bdraco committed:
Increase discovery frequency during migration

bdraco commented:
I increased the discovery interval to every 5m instead of 15m when the migration is happening so it should complete a bit faster.  Once the discovery finishes it will go back to 15m

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
lint

bdraco commented:
So it does look like the upstream lib can block longer than the time so we also need to handle cancelation on timeout.


```
2022-07-07 16:51:01.602 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 214, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/aiolifx/aiolifx.py", line 306, in try_sending
    myresult = await aio.wait_for(event.wait(), timeout_secs)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/aiolifx/aiolifx.py", line 313, in try_sending
    callb(self, None)
  File "/usr/src/homeassistant/homeassistant/components/lifx/util.py", line 148, in _callback
    future.set_result(message)
asyncio.exceptions.InvalidStateError: invalid state
```

bdraco committed:
do not throw on cancelation

bdraco committed:
typo in comment

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
mock number of broadcast interfaces so its consistant between runners

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
adjust

bdraco committed:
Merge branch 'lifx_migrate_config_entry_per_device' of github.com:bdraco/home-assistant into lifx_migrate_config_entry_per_device
ase enter a commit message to explain why this merge is necessary,

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco commented:
Updating only to ensure everything still works with latest dev

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bdraco committed:
Merge branch 'dev' into lifx_migrate_config_entry_per_device

bachya commented on homeassistant/components/lifx/__init__.py:
```diff
@@ -22,39 +45,173 @@
+MIGRATION_INTERVAL = timedelta(minutes=5)
+
+DISCOVERY_COOLDOWN = 5
+_LOGGER = logging.getLogger(__name__)
```
(Maybe in a future PR) This can go as a single `LOGGER` in `const.py` (which will free up each platform from having to define its own).

bdraco replied:
c90f5e1f34c8dffde5437f81e800c01f9b8dd714

bachya commented on homeassistant/components/lifx/__init__.py:
```diff
@@ -22,39 +45,173 @@
+    # to reduce the risk we miss devices because the event
+    # loop is blocked at startup.
+    discovery_manager.async_setup_discovery_interval()
+    async_call_later(hass, DISCOVERY_COOLDOWN, _async_delayed_discovery)
```
Why not just wait for `EVENT_HOMEASSISTANT_STARTED` and do discovery then? Would avoid the cooldown need?

bdraco replied:
We want to try to migrate them before the started even if possible so we don't have the least amount of devices that churn after startup

bachya commented on homeassistant/components/lifx/__init__.py:
```diff
@@ -22,39 +45,173 @@
+    connection = LIFXConnection(host, TARGET_ANY)
+    try:
+        await connection.async_setup()
+    except socket.gaierror as ex:
```
I assume `aiolifx` doesn't wrap this in a LIFX-specific error?

bdraco replied:
This happens when the user ends a dns name and it cannot be resolved.  The lib doesn't handle this

bachya commented on homeassistant/components/lifx/config_flow.py:
```diff
@@ -1,16 +1,242 @@
+        try:
+            await connection.async_setup()
+        except socket.gaierror:
+            return None
```
If this can return None, we should update this function's return type: `Light | None`

bdraco replied:
aa986159711fe18fadf59d97954f78789afebdbe

bachya commented on homeassistant/components/lifx/config_flow.py:
```diff
@@ -1,16 +1,242 @@
+            for progress in self._async_in_progress()
+        ):
+            return self.async_abort(reason="already_in_progress")
+        device = await self._async_try_connect(
```
```python
        if not (
            device := await self._async_try_connect(
                host, serial=serial, raise_on_progress=True
            )
        ):
            return self.async_abort(reason="cannot_connect")
```

bdraco replied:
6007e963e008b935eaaa54e4b6ae6d5774a71cec

bachya reviewed (commented):
I don't have LIFX devices to test, so I'll assume functional testing is good to go elsewhere.

bdraco committed:
fix typing

bdraco committed:
tweak

bdraco committed:
move DATA_LIFX_MANAGER

bdraco commented on homeassistant/components/lifx/__init__.py:
```diff
@@ -22,39 +45,173 @@
+    assert entry.unique_id is not None
+    if DATA_LIFX_MANAGER not in hass.data:
+        manager = LIFXManager(hass)
+        hass.data[DATA_LIFX_MANAGER] = manager
```
85bc82f0c49badd4c50216837378587417b29473

bdraco committed:
loggers

bdraco reviewed (commented):
Addressed review comments, and left two notes

bdraco commented:
Retesting now

bdraco committed:
fixes

bdraco committed:
tweak

bdraco commented:
Functional retest

- [x] migration
- [x] discovery
- [x] setup
- [x] color
- [ ] effects

Tests are passing

bachya reviewed (approved):
LGTM 👍🏻

Djelibeybi commented:
@bdraco this is a gentle reminder to please create the skeleton of a diagnostic sensor so I can start adding things like RSSI, and Infrared/HEV status/config, etc. :)

bdraco commented:
@Djelibeybi Do you have an example of fetching the RSSI with `aiolifx` ?

bdraco commented:
Also since this is merged, we should move the discussion to discord `bdraco#6479`

Djelibeybi commented:
I'm at work but I'll try and get onto Discord later.

MartinHjelmare commented on homeassistant/components/lifx/migration.py:
```diff
@@ -0,0 +1,74 @@
+from .discovery import async_init_discovery_flow
+
+
+async def async_migrate_legacy_entries(
```
These functions don't need to be coroutine functions since we don't await inside.

bdraco replied:
https://github.com/home-assistant/core/pull/75719

👍

MartinHjelmare commented on homeassistant/components/lifx/util.py:
```diff
@@ -0,0 +1,161 @@
+
+async def async_execute_lifx(method: Callable) -> Message:
+    """Execute a lifx coroutine and wait for a response."""
+    future: asyncio.Future[Message] = asyncio.Future()
```
Would be nice to implement a simpler interface in the library so the user can just await the library methods.

bdraco replied:
I agree.  There are a few upstream merges waiting though so it might be a while.

bdraco replied:
Queued for when upstream is able to review PRs again (there are multiple waiting)
